dataset/src/main/scala/frameless/ops/MatchesAll.scala

```diff
@@ -0,0 +1,19 @@
+package frameless
+package ops
+
+import shapeless.HList
+import shapeless.ops.hlist.Filter
+
+import scala.annotation.implicitNotFound
+
+
+@implicitNotFound("HList of type '${L}' is NOT comprised of only elements of type '${U}'")
+trait MatchesAll[L <: HList, U]
+
+object MatchesAll {
+  implicit def matchesAll[L <: HList, U, Rest <: HList]
+  (implicit
+   filtered: Filter.Aux[L, U, Rest],
+   remainedSame: Rest =:= L
```

Comment:
Actually you probably if `MatchesAll` is equivalant to a `FilterNot` you probably don't need a new type class at all